### DISCUSSION TITLE
Invalid window id Issue with NuiSplit

### DISCUSSION THREAD

**Discussion by MunifTanjim** [MunifTanjim/nui.nvim#335] [Q&A]
Possibly one thing;

Suppose there are two windows (`1` and `2`) side by side. When I make a new `NuiSplit({ relative = "editor" }):mount()` (creates `3`), it opens at the most left(/right) position.
Inside `:update_layout()` inside `:init()`, `self._.relative` is changed to `{ type = "editor", win = 1 }`.

After that, I `split:hide()` and then delete window `1`.

When I call `split:show()` after that, the value in `self._.relative.win` is required [here](https://github.com/MunifTanjim/nui.nvim/blob/af7dfee12fbf51d12cfc6ee386fa54f7a5a573c8/lua/nui/split/init.lua#L168), and since window `1` is already deleted, `:show()` will fail.

To workaround this issue, I need to call `split:update_layout()` before `split:show()` to ensure `self._.relative.win` is a valid window.

When I'm making a split relative to the `editor`, I suppose the code shouldn't matter where to spawn the window from, so it would be more intuitive if I didn't need to call update_layout. (And it took quite a lot of time to figure out...)

It would be nice if you could add some if-else based on relative settings or possibly put a caution in the document?

_orignally posted by @pysan3 in https://github.com/MunifTanjim/nui.nvim/pull/295#issuecomment-1945403149_

**Comment by MunifTanjim** [comment]
@pysan3 Can you try changing this https://github.com/MunifTanjim/nui.nvim/blob/35da9ca1de0fc4dda96c2e214d93d363c145f418/lua/nui/split/init.lua#L168-L168

to

```lua
  self.winid = vim.api.nvim_win_call(self._.relative.type == 'editor' and 0 or self._.relative.win, function()
```

and see if that works?

**Reply by pysan3** [reply]
Yep ;)

Ideally, we would like to keep `self._.relative.win` to 0 while `self._.relative.type == "editor"` which is modified here.

https://github.com/MunifTanjim/nui.nvim/blob/af7dfee12fbf51d12cfc6ee386fa54f7a5a573c8/lua/nui/split/utils.lua#L132-L152

But I'm not 100% sure if `self._.relative.win == 0` does cause issues / make breaking changes to existing code.

NuiPopup is dealt here in a different place btw.

https://github.com/MunifTanjim/nui.nvim/blob/af7dfee12fbf51d12cfc6ee386fa54f7a5a573c8/lua/nui/layout/utils.lua#L137-L147

**Reply by pysan3** [reply]
Seems like we do need this fixed as I still experience invalid window error with noice.

@MunifTanjim 

> ```lua
>   self.winid = vim.api.nvim_win_call(self._.relative.type == 'editor' and 0 or self._.relative.win, function()
> ```

I think your suggestion is good enough.

And looking forward to seeing a new version release as well :)

**Reply by MunifTanjim** [reply]
Sorry, I couldn't find the time to fix this. If you open a PR I'd be happy to review and merge it.

**Reply by pysan3** [reply]
https://github.com/MunifTanjim/nui.nvim/pull/337

**Comment by pysan3** [comment]
May I ask you why you reference `relative.winid` here?

https://github.com/MunifTanjim/nui.nvim/blob/af7dfee12fbf51d12cfc6ee386fa54f7a5a573c8/lua/nui/split/utils.lua#L68

Called from

https://github.com/MunifTanjim/nui.nvim/blob/af7dfee12fbf51d12cfc6ee386fa54f7a5a573c8/lua/nui/split/utils.lua#L142

If I'm not mistaken, `options.win` the only thing user can pass according to the type annotations. Or am I missing something?

**Reply by MunifTanjim** [reply]
> options.win the only thing user can pass according to the type annotations

Not sure which type annotation you're looking at. But for split, this is it:

https://github.com/MunifTanjim/nui.nvim/blob/af7dfee12fbf51d12cfc6ee386fa54f7a5a573c8/lua/nui/split/init.lua#L99-L106

**Reply by pysan3** [reply]
I meant `options.relative.win` where the code in `split/utils.lua` (linked above) refers to `options.relative.winid`.

`nui_split_option_relative` has only `type: 'editor'|'win', win: number` right? No `winid` field.

**Reply by MunifTanjim** [reply]
Ah, I see! 🤔 It might be a typo.

**Reply by pysan3** [reply]
But that code has been there for more than an year, and I'm not brave enough to blindly fix it lol.

I'll hand it to you to decide and fix the problem.

**Reply by MunifTanjim** [reply]
Okay so the type annotation is wrong. This is the doc:

https://github.com/MunifTanjim/nui.nvim/blob/af8ddf5db7e8485051aacefb24d76ab24ea26a0c/lua/nui/split/README.md?plain=1#L46-L53

**Reply by pysan3** [reply]
Ah, that's why my hand always typed winid there, although my brain told me to write win.

My hands remembered the old era with no type annotations lol.

**Reply by MunifTanjim** [reply]
Should be fixed now.

**Reply by pysan3** [reply]
@MunifTanjim 
https://github.com/MunifTanjim/nui.nvim/commit/7da355e699d15317f0600bbd2bccead7e9d97e80#commitcomment-138737872
